### PR TITLE
Wire up Error Report Http Client

### PR DESCRIPTION
**Pull request by DanVanAtta**
## Overview
- Wire up the error report upload client
  - Add configuration to instantiate the client
  - Add a 'handler' class "ErrorReportUploadAction" that will invoke the
    client and show a progress spinner while the upload is in-progress.
    The 'handler' is responsible for looking at the result and showing
    a success or error message.
- Add a user setting option to override http server URI
- Add convenience method to BackGroundTaskRunner that does not throw
InterruptedException

## Screeshot
screenshot from 2019-01-04 02-12-30

## Functional Changes
- Sending reports with the send window will now do something and try to send the error report to our http server

## Manual Testing Performed
- turned on test-feature flag in settings, sent some reports. They failed, but they failed on the server and not due to any client issue.

## Reviewer Notes
Still quite a bit to do as follow-up:
- resolve server issues that prevent upload
- do a better job of wiring server error message back to client
- keep improving error handling on client side
- client side validation to disallow empty data from being sent

### CODE REVIEW

File: game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+class ErrorReportConfiguration {
+
+  /**
+   * Creates a 'report handler', this is the component invoked when we have collected all data from
+   * user and wish to then send that data. The report handler is responsible for showing a waiting
+   * dialog and sending error report data to the remote http server.
+   */
+  static BiConsumer<JFrame, UserErrorReport> newReportHandler() {
+    if (ClientSetting.httpLobbyUriOverride.isSet()) {
```
Comment (ssoloff) [inline game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportConfiguration.java]: Might be more idiomatic here to use `getValue()` and `map()` and `orElseGet()`:

```java
return ClientSetting.httpLobbyUriOverride.getValue()
    .<BiConsumer<JFrame, UserErrorReport>>map(
        httpLobbyUri -> new ErrorReportUploadAction(
            ErrorReportClientFactory.newErrorUploader(
                URI.create(httpLobbyUri))))
    .orElseGet(() -> LobbyPropertyFetcherConfiguration.lobbyServerPropertiesFetcher()
        .fetchLobbyServerProperties()
        .<BiConsumer<JFrame, UserErrorReport>>map(
            lobbyServerProperties -> new ErrorReportUploadAction(
                ErrorReportClientFactory.newErrorUploader(
                    lobbyServerProperties.getHttpServerUri())))
        .orElse(ErrorReportUploadAction.OFFLINE_STRATEGY));
```

File: game-core/src/main/java/games/strategy/triplea/settings/ClientSettingSwingUiBinding.java [hunk 2]
```diff
@@ -228,6 +228,16 @@
     }
   },
 
+  TEST_HTTP_LOBBY_URI(
+      "Http Lobby URI",
```
Comment (ssoloff) [inline game-core/src/main/java/games/strategy/triplea/settings/ClientSettingSwingUiBinding.java]: We should probably consistently display acronyms to the user.  In this one block, we have three different styles: `Http`, `URI`, and `http`.

File: game-core/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java [hunk 3]
```diff
@@ -52,6 +53,13 @@ public void runInBackground(
     });
   }
 
+  /**
+   * Similar to 'runInBackground' except uses Interruptibles#await to suppress the {@code InterruptedException}.
+   */
+  public void awaitRunInBackground(final String message, final Runnable backgroundAction) {
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java]: The primary motivator is to remove the repeated `Interruptibles` wrapping. 
EG:
```
  Interruptibles.await(() -> GameRunner.newBackgroundTaskRunner().runInBackground(
```
To:
```
GameRunner.newBackgroundTaskRunner().runInBackground(
```

File: game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportConfiguration.java (hunk 1 above)
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportConfiguration.java]: Thanks for poiting this out, I was able to run with the idea and added: 9df2a737e

The same chunk of code now looks like this:
```
  static BiConsumer<JFrame, UserErrorReport> newReportHandler() {
    return newFromUserOverride()
        .orElseGet(
            () -> newFromRemoteProperties().orElse(ErrorReportUploadAction.OFFLINE_STRATEGY));
  }

  private static Optional<BiConsumer<JFrame, UserErrorReport>> newFromUserOverride() {
    return ClientSetting.httpLobbyUriOverride
        .getValue()
        .map(URI::create)
        .map(ErrorReportClientFactory::newErrorUploader)
        .map(ErrorReportUploadAction::new);
  }

  private static Optional<BiConsumer<JFrame, UserErrorReport>> newFromRemoteProperties() {
    return LobbyPropertyFetcherConfiguration.lobbyServerPropertiesFetcher()
        .fetchLobbyServerProperties()
        .map(
            lobbyServerProperties ->
                new ErrorReportUploadAction(
                    ErrorReportClientFactory.newErrorUploader(
                        lobbyServerProperties.getHttpServerUri())));
  }
```

File: game-core/src/main/java/games/strategy/triplea/settings/ClientSettingSwingUiBinding.java (hunk 2 above)
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/triplea/settings/ClientSettingSwingUiBinding.java]: Easy enough to fix: ac2395383

Side-topic: I somewhat wonder if we should develop a code name for the http lobby, so we can call it by something else? The old lobby might be well named 'sanjuro' to match one of its earliest names.

File: game-core/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ssoloff) [reply game-core/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java]: It just seems like a bit of mixing unrelated concerns.

The problem we had previously was a lot of `InterruptedException`s were being buried within code, which made it hard to later run that code "interruptibly" in a background thread.  Going down this route may once again promote that kind of code structure.

The idea behind extracting the `Interruptibles` class was for it to be the single source of the smell of burying `InterruptedException`.  If that makes sense. :man_shrugging:

File: game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportConfiguration.java (hunk 1 above)
Comment (ssoloff) [reply game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportConfiguration.java]: :+1:

File: game-core/src/main/java/games/strategy/triplea/settings/ClientSettingSwingUiBinding.java (hunk 2 above)
Comment (ssoloff) [reply game-core/src/main/java/games/strategy/triplea/settings/ClientSettingSwingUiBinding.java]: Just as long as the code names have an A&A flair to them. :smile:

**Pull request merged**